### ISSUE TITLE
Critical security issue

### ISSUE THREAD

**Issue by DCNick3** [Aimeast/FxSsh#22]
Result of signature verification in public key auth is ignored. Because of that server will accept any signature effectively removing need of private key for auth. This allows user to auth using ONLY trusted public key.

**Comment by Aimeast** [comment]
In the case which I have tested, the client (ssh.exe) may working on lazy mode of exchange message.
Here is the message sequence diagram that I record at Jun, 2016. 
![sequence diagram](https://static.swimlanes.io/cf118228ebab91171c69c26ade017ab0.png)
Sequence diagram image short link as https://swimlanes.io/d/MjrWryY19

If you send a `SSH_MSG_USERAUTH_FAILURE` to client before seq. 4, you may lose the connection.

The seq. 5, server side will verify the signature and send `SSH_MSG_USERAUTH_FAILURE` if not verified. 

It is true there has critical security issue, but it not simple as patch #23. If you remove the line 96 `e.Result = true` in `service_Userauth`, you may thought it out. The `Userauth?.Invoke(this, args)` at wrong place.

**Comment by DCNick3** [comment]
The `SSH_MSG_USERAUTH_REQUEST` without signature is handled earlier, at [L95-L99](https://github.com/DCNick3/FxSsh/blob/089ac0d0dd34d869b8fc8f19bd37cccf226ca063/FxSsh/Services/UserauthService.cs#L95-L99)

Can't see why `Userauth?.Invoke(this, args)` is at wrong place too. It checks whether the server trusts owner of private key for specified private key, and this is done before checking any signature, as specified by RFC 4252:
>  When the server receives this message, it MUST check whether the
   supplied key is acceptable for authentication, and if so, it MUST
   check whether the signature is correct.

**Comment by DCNick3** [comment]
Any thoughts on it?

**Comment by Aimeast** [comment]
Sorry, I am very busy recently. Tomorrow maybe free.

**Comment by Aimeast** [comment]
I have make a patch, do you want a try to check and more verify.

```patch
diff --git a/FxSsh/Services/UserauthArgs.cs b/FxSsh/Services/UserauthArgs.cs
index 6665461..b72b091 100644
--- a/FxSsh/Services/UserauthArgs.cs
+++ b/FxSsh/Services/UserauthArgs.cs
@@ -4,7 +4,7 @@ namespace FxSsh.Services
 {
     public class UserauthArgs
     {
-        public UserauthArgs(Session session, string username, string keyAlgorithm, string fingerprint, byte[] key)
+        public UserauthArgs(Session session, string username, string keyAlgorithm, string fingerprint, byte[] key, bool signatureVerified)
         {
             Contract.Requires(keyAlgorithm != null);
             Contract.Requires(fingerprint != null);
@@ -14,6 +14,7 @@ namespace FxSsh.Services
             KeyAlgorithm = keyAlgorithm;
             Fingerprint = fingerprint;
             Key = key;
+            SignatureVerified = signatureVerified;
             Session = session;
             Username = username;
         }
@@ -36,6 +37,13 @@ namespace FxSsh.Services
         public string KeyAlgorithm { get; private set; }
         public string Fingerprint { get; private set; }
         public byte[] Key { get; private set; }
-        public bool Result { get; set; }
+        public bool SignatureVerified { get; private set; }
+        /// <summary>
+        /// A identity of authorition result.
+        /// True, for succeed.
+        /// False, for NOT succeed.
+        /// Null, for undetermined.
+        /// </summary>
+        public bool? Result { get; set; }
     }
 }
diff --git a/FxSsh/Services/UserauthService.cs b/FxSsh/Services/UserauthService.cs
index 718a220..f91e96e 100644
--- a/FxSsh/Services/UserauthService.cs
+++ b/FxSsh/Services/UserauthService.cs
@@ -49,23 +49,14 @@ namespace FxSsh.Services
 
         private void HandleMessage(PasswordRequestMessage message)
         {
-            var verifed = false;
-
             var args = new UserauthArgs(_session, message.Username, message.Password);
-            if (Userauth != null)
-            {
-                Userauth(this, args);
-                verifed = args.Result;
-            }
+            Userauth?.Invoke(this, args);
 
-            if (verifed)
+            if (args.Result == true)
             {
                 _session.RegisterService(message.ServiceName, args);
-
                 Succeed?.Invoke(this, message.ServiceName);
-
                 _session.SendMessage(new SuccessMessage());
-                return;
             }
             else
             {
@@ -77,40 +68,40 @@ namespace FxSsh.Services
         {
             if (Session._publicKeyAlgorithms.ContainsKey(message.KeyAlgorithmName))
             {
-                var verifed = false;
-
                 var keyAlg = Session._publicKeyAlgorithms[message.KeyAlgorithmName](null);
                 keyAlg.LoadKeyAndCertificatesData(message.PublicKey);
 
-                var args = new UserauthArgs(base._session, message.Username, message.KeyAlgorithmName, keyAlg.GetFingerprint(), message.PublicKey);
+                var rightSig = false;
+                if (message.HasSignature)
+                {
+                    var sig = keyAlg.GetSignature(message.Signature);
+
+                    using (var worker = new SshDataWorker())
+                    {
+                        worker.WriteBinary(_session.SessionId);
+                        worker.Write(message.PayloadWithoutSignature);
+
+                        rightSig = keyAlg.VerifyData(worker.ToByteArray(), sig);
+                    }
+                }
+
+                var args = new UserauthArgs(_session, message.Username, message.KeyAlgorithmName, keyAlg.GetFingerprint(), message.PublicKey, rightSig);
                 Userauth?.Invoke(this, args);
-                verifed = args.Result;
 
-                if (!verifed)
+                if (args.Result == true)
                 {
-                    _session.SendMessage(new FailureMessage());
-                    return;
+                    _session.RegisterService(message.ServiceName, args);
+                    Succeed?.Invoke(this, message.ServiceName);
+                    _session.SendMessage(new SuccessMessage());
                 }
-
-                if (!message.HasSignature)
+                else if (args.Result == null && !message.HasSignature)
                 {
                     _session.SendMessage(new PublicKeyOkMessage { KeyAlgorithmName = message.KeyAlgorithmName, PublicKey = message.PublicKey });
-                    return;
                 }
-
-                var sig = keyAlg.GetSignature(message.Signature);
-
-                using (var worker = new SshDataWorker())
+                else
                 {
-                    worker.WriteBinary(_session.SessionId);
-                    worker.Write(message.PayloadWithoutSignature);
-
-                    verifed = keyAlg.VerifyData(worker.ToByteArray(), sig);
+                    _session.SendMessage(new FailureMessage());
                 }
-
-                _session.RegisterService(message.ServiceName, args);
-                Succeed?.Invoke(this, message.ServiceName);
-                _session.SendMessage(new SuccessMessage());
             }
         }
     }
diff --git a/SshServerLoader/Program.cs b/SshServerLoader/Program.cs
index d8495b5..c460b24 100644
--- a/SshServerLoader/Program.cs
+++ b/SshServerLoader/Program.cs
@@ -91,9 +91,20 @@ namespace SshServerLoader
 
         static void service_Userauth(object sender, UserauthArgs e)
         {
-            Console.WriteLine("Client {0} fingerprint: {1}.", e.KeyAlgorithm, e.Fingerprint);
+            Console.WriteLine($"Client auth method {e.AuthMethod}, username {e.Username}, key {e.KeyAlgorithm}, fingerprint: {e.Fingerprint}.");
 
-            e.Result = true;
+            if (e.AuthMethod == "password")
+            {
+                e.Result = true;  // func(e.Username, e.Password);
+            }
+            else if (e.AuthMethod == "publickey")
+            {
+                if (e.SignatureVerified)
+                {
+                    e.Result = true;  // func(e.Fingerprint);
+                }
+                // e.Result = null;  // if no signature.
+            }
         }
 
         static void service_CommandOpened(object sender, CommandRequestedArgs e)
```

**Comment by DCNick3** [comment]
I still don't understand why do you need such a thing. If signature was sent it must be correct. Openssh responds with `SSH_MSG_AUTH_FAILURE`. 

To check if server trusts certain key without having access to private part there is form of message without signature. Your implementation currently handles it right by responding with `SSH_MSG_USERAUTH_PK_OK`.

So, what is the purpose of allowing auth with invalid signature specified (possible for buggy clients or sup3r h4xors)?

**Comment by Aimeast** [comment]
Document [rfc4252 SSH Authentication Protocol @ 7.  Public Key Authentication Method: "publickey"](https://tools.ietf.org/html/rfc4252#section-7) say

> To perform actual authentication, the client MAY then send a signature generated using the private key.  The client MAY send the signature directly without first verifying whether the key is acceptable.

and
> The server MUST respond to this message with either SSH_MSG_USERAUTH_FAILURE **or with the following**: SSH_MSG_USERAUTH_PK_OK


The behavior of two clients tested:
1. putty: send key and signature in first `SSH_MSG_USERAUTH_REQUEST`
2. ssh.exe(on git for windows): send key and WITHOUT signature in first `SSH_MSG_USERAUTH_REQUEST`, only send signature after received `SSH_MSG_USERAUTH_PK_OK`. connection will be lost if received a `SSH_MSG_AUTH_FAILURE`.

Difference clients have difference behaviors. Even more, send `SSH_MSG_USERAUTH_PK_OK` from server side is not wrong. The server side user (I mean the administrator on program.cs) will decide it is accepted or not, I think.

**Comment by DCNick3** [comment]
Yes, this is totally ok. But `SSH_MSG_USERAUTH_REQUEST` without signature is not the same as `SSH_MSG_USERAUTH_REQUEST` with invalid signature

Here's form without signature:
```
      byte      SSH_MSG_USERAUTH_REQUEST
      string    user name in ISO-10646 UTF-8 encoding [RFC3629]
      string    service name in US-ASCII
      string    "publickey"
      boolean   FALSE
      string    public key algorithm name
      string    public key blob
```

The false corresponds to `HasSignature` property in your implementation. Notice, there is no signature field. This case is correctly handled [here](https://github.com/DCNick3/FxSsh/blob/089ac0d0dd34d869b8fc8f19bd37cccf226ca063/FxSsh/Services/UserauthService.cs#L95-L99). If server trusts provided public key it sends `SSH_MSG_USERAUTH_PK_OK`. After this openssh will get access to private key (maybe with asking user for password to decrypt it) and make another form of `SSH_MSG_USERAUTH_REQUEST` with signature:
```
      byte      SSH_MSG_USERAUTH_REQUEST
      string    user name
      string    service name
      string    "publickey"
      boolean   TRUE
      string    public key algorithm name
      string    public key to be used for authentication
      string    signature
```

> When the server receives this message, it MUST check whether the
> supplied key is acceptable for authentication, and if so, it MUST
> check whether the signature is correct.
>
> If both checks succeed, this method is successful.

So, if client sends signature right away or after receiving `SSH_MSG_USERAUTH_PK_OK` the signature MUST be correct. If them wants to probe some public key to know whether server trusts it, them will send a `SSH_MSG_USERAUTH_REQUEST` with `HasSignature` set to false, so no signature will be sent. Invalid signature is COMPLETELY WRONG  and it indicates error in protocol.

The point I have: your current implementation is almost correct. The problem is that if client supplies invalid signature to server (which, again, is a protocol error), it will be treated same as when the signature is valid. 

As I tested my ssh client implementation based on your project I accidentally messed up signing and sent `SSH_MSG_USERAUTH_REQUEST` with invalid signature. Openssh (the project where ssh.exe from git for windows comes too) written `crypto_sign_ed25519_open failed: -1`  to log (well, it means, that it can't verify signature) and send a `SSH_MSG_USERAUTH_FAILURE`.

**Comment by Aimeast** [comment]
I got your point. Looks like should be this
```cs
using (var worker = new SshDataWorker())
{
    worker.WriteBinary(_session.SessionId);
    worker.Write(message.PayloadWithoutSignature);

    var verified = keyAlg.VerifyData(worker.ToByteArray(), sig);
    if (!verified)
    {
        _session.SendMessage(new FailureMessage());
        return;
    }
}
```

**Comment by DCNick3** [comment]
Yep, this looks like it. I have offered very similar patch if I am not mistaken.

**Comment by Aimeast** [comment]
OK, go ahead, update the PR.

**Comment by DCNick3** [comment]
As far as I can tell it is practically equivalent. Two differences: it doesn't introduce new verified variable and performs it's check out of using scope:

```c#
using (var worker = new SshDataWorker())
{
    worker.WriteBinary(_session.SessionId);
    worker.Write(message.PayloadWithoutSignature);

    verifed = keyAlg.VerifyData(worker.ToByteArray(), sig);
}

if (!verifed)
{
    _session.SendMessage(new FailureMessage());
    return;
}
```

Or should I change something?

**Comment by Aimeast** [comment]
Please do it. The goal is good code, good performance

**Comment by DCNick3** [comment]
The main question is: what should I change? I don't really see any significant differences in your and my code, so please state what should be changed.

**Comment by Aimeast** [comment]
`Userauth?.Invoke(this, args)` at wrong place. The original code `Userauth?.Invoke(this, args)` before verify signature, and `Userauth` event can't got the status which without signature.

**Comment by DCNick3** [comment]
As far as I can tell the whole point of `Userauth` event is to `check whether the
supplied key is acceptable for authentication` (1st part of auth check according to RFC). After that code in `HandleMessage` checks the signature. If message would be sent without the signature the event would still be invoked to check if server trusts the key to tell the client if them should send signature.

**Comment by Aimeast** [comment]
The process logic should be:

1. check has signature, require signature if no. by `SSH_MSG_USERAUTH_PK_OK`
2. verify signature, fail if not verified. by `SSH_MSG_USERAUTH_FAILURE`
3. invoke user auth event, send success or failure. by `SSH_MSG_USERAUTH_SUCCESS` or `SSH_MSG_USERAUTH_FAILURE`

user auth event should not consider the signature right or not. if invoke user auth event, the signature should be verified inside.

**Comment by DCNick3** [comment]
RFC4252 says:
> Private keys are often stored in an encrypted form at the client
>   host, and the user must supply a passphrase before the signature can
>   be generated.  Even if they are not, the signing operation involves
>   some expensive computation.  To avoid unnecessary processing and user
>   interaction, the following message is provided for querying whether
>   authentication using the "publickey" method would be acceptable.

So, the whole point of sending the message without signature is to check if server will accept such a key. If server will always send `SSH_MSG_USERAUTH_PK_OK` this idea will lose all it's purpose. Also, RFC says to verify whether key is trusted before the signature verification:
>When the server receives this message, it MUST check whether the
>   supplied key is acceptable for authentication, and if so, it MUST
>   check whether the signature is correct.

**Comment by Aimeast** [comment]
Yeah, your explain perfectly clear. And reviewed PR again, it is the same as the explain. I accept it.

The only question for now is the design will invoke `UserAuth` twice if first request without signature. How about design a mechanism which invoke auth only once?

**Comment by DCNick3** [comment]
Is there any problem with that? AFAIK it is already invoked twice... Can't think of any method to do this without breaking RFC accordance. Still, what problems can this introduce?

**Comment by Aimeast** [comment]
It is a new question not scope of this issue, I will consider it later. Please make PR rebase to new dev. Thanks for your patience explain, Thank you very much!

**Comment by DCNick3** [comment]
What do you mean by "rebase to new dev"? It is already targeted to `dev` branch and squashed to one commit...

**Comment by Aimeast** [comment]
eennnn, since last PR merged, the base was changed to e5170e2, not dbacd36. Just `git rebase e5170e2`. I mean, keep comments as a line.

**Comment by DCNick3** [comment]
Done.
